src/ports/SkMemory_info.cpp

```diff
@@ -0,0 +1,56 @@
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <assert.h>
+#include <emscripten/emscripten.h>
+#include "include/private/SkMalloc.h"
+
+struct s_mallinfo {
+    int arena;    /* non-mmapped space allocated from system */
+    int ordblks;  /* number of free chunks */
+    int smblks;   /* always 0 */
+    int hblks;    /* always 0 */
+    int hblkhd;   /* space in mmapped regions */
+    int usmblks;  /* maximum total allocated space */
+    int fsmblks;  /* always 0 */
+    int uordblks; /* total allocated space */
+    int fordblks; /* total free space */
+    int keepcost; /* releasable (via malloc_trim) space */
+};
+
+extern "C" {
+    extern s_mallinfo mallinfo();
+}
+
+/**
+ * Emscripten uses:
+ * wasmMemory = new WebAssembly.Memory({ 'initial': ... , 'maximum': ...})
+ * Unfortunately this buffer as well as its parameters are not directly
+ * accessible, hence providing some extra API.
+ */
+unsigned int sk_get_initial_memory() {
+    return SK_INITIAL_MEMORY_MB * 1024 * 1024;
+}
+
+unsigned int sk_get_maximum_memory() {
+    return SK_MAXIMUM_MEMORY_MB * 1024 * 1024;
+}
+
+unsigned int sk_get_total_memory() {
+    return EM_ASM_INT(return HEAP8.length);
+}
+
+/**
+ * Returns available heap size in bytes. Probably not too fragmented as it
+ * is allocated on WebAssembly.Memory(...). Yet this number is relative to
+ * the currently available getTotalMemory() heap. Thus:
+ *  currentHeapAllocated = sk_get_total_memory() - sk_get_free_memory()
+ *  maximalFreeMemory = sk_get_maximum_memory() - (sk_get_total_memory() - sk_get_free_memory())
+ */
+unsigned int sk_get_free_memory() {
+    s_mallinfo info = mallinfo();
+    unsigned int totalMemory = sk_get_total_memory();
+    unsigned int dynamicTop = (unsigned int)sbrk(0);
+    return totalMemory - dynamicTop + info.fordblks;
```

Comment:
okay, this is obviously the whole point of the PR.

I think there's a "units problem" here.  `totalMemory` and `info.fordblks` are counts of memory.  But
`dynamicTop = sbrk(0)` returns a memory address -- I don't understand how that fits with the counts.

Reply:
I understand that this is in the WASM environment, so there could be some "monkey business" wrt to addresses.  But I'm still unclear about the implications of that.

Reply:
From above: 
 * Emscripten uses:
 * wasmMemory = new WebAssembly.Memory({ 'initial': ... , 'maximum': ...})
 * Unfortunately this buffer as well as its parameters are not directly
 * accessible, hence providing some extra API.

@ronpfeifle it follows that we have all our HEAP8 as a buffer with adresses of:
0 ... HEAP8.length
Hence to get the current chunk one needs to:
totalMemory - dynamicTop

Reply:
Ah -- okay, that's just the way WASM works, I guess.  Thanks for the explanation! 😄